modules/carbonAnalyticsAdapter.js

```diff
@@ -0,0 +1,238 @@
+  }
+
+  pageViewId = generateUUID();
+  profileId = getProfileId();
+  sessionId = getSessionId();
+
+  pageEngagement = { // create the initial page engagement event
+    ttl: 60, // unit is seconds
+    count: 0,
+    id: generateUUID(),
+    timeLastEngage: Date.now()
+  };
+
+  let event = createBaseEngagementEvent()
+  sendEngagementEvent(event, 'page_load');
+
+  carbonAdapter.originEnableAnalytics(config); // call the base class function
+};
+
+function getProfileId() {
+  if (storage.localStorageIsEnabled()) {
+    let localStorageId = storage.getDataFromLocalStorage(PROFILE_ID_KEY);
+    if (localStorageId && localStorageId != '') {
+      if (storage.cookiesAreEnabled()) {
+        storage.setCookie(PROFILE_ID_COOKIE, localStorageId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+      }
+
+      return localStorageId;
+    }
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(PROFILE_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      if (storage.localStorageIsEnabled()) {
+        storage.setDataInLocalStorage(PROFILE_ID_KEY, cookieId);
+      }
+
+      storage.setCookie(PROFILE_ID_COOKIE, cookieId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+
+      return cookieId;
+    }
+  }
+
+  let newId = generateUUID();
+  if (storage.localStorageIsEnabled()) {
+    storage.setDataInLocalStorage(PROFILE_ID_KEY, newId);
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    storage.setCookie(PROFILE_ID_COOKIE, newId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+  }
+
+  return newId;
+}
+
+function getSessionId() {
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(SESSION_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      storage.setCookie(SESSION_ID_COOKIE, cookieId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+      return cookieId;
+    }
+
+    let newId = generateUUID();
+    storage.setCookie(SESSION_ID_COOKIE, newId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+    return newId;
+  }
+
+  return generateUUID();
+}
+
+function registerEngagement() {
+  let present = Date.now();
+  let timediff = (present - pageEngagement.timeLastEngage) / 1000; // convert to seconds
+
+  pageEngagement.timeLastEngage = present;
+
+  if (timediff < pageEngagement.ttl) {
+    return;
+  }
+
+  pageEngagement.count++;
+  pageEngagement.id = generateUUID();
+};
+
+function getConsentData(args) {
+  let consentData = {
+    gdpr_consent: '',
+    ccpa_consent: ''
+  };
+
+  if (Array.isArray(args?.bidderRequests) && args.bidderRequests.length > 0) {
+    let bidderRequest = args.bidderRequests[0];
+
+    if (bidderRequest?.gdprConsent?.consentString) {
+      consentData.gdpr_consent = bidderRequest.gdprConsent.consentString;
+    }
+
+    if (bidderRequest?.uspConsent?.consentString) {
+      consentData.ccpa_consent = bidderRequest.uspConsent.consentString;
+    }
+  }
+
+  return consentData;
+}
+
+function getExternalIds() {
+  let externalIds = {};
```

Comment:
If there are no external ids this still sends as empty object.

Is this intended?

If not, just return inside the below IF and let it return null or undefined otherwise and it should remove it when it is sent.